### PR TITLE
Conformance tests for migration-Pt.1

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
This PR was split into 2 parts after the review round. (Pt.2: https://github.com/kubevirt/kubevirt/pull/13066)

This PR aims to identify and label the tests in the migration pkg.
Also, a cleanup and refactoring of the tests is performed:
- convert e2e test into a unit test if possible;
- remove tests that do not add anything to an already present unit test;
- use patchset;
- remove duplicates;
- merge similar tests.

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by fossedihelm** [comment]
/hold
Need to recheck some conformance tests

File: tests/migration/eviction_strategy.go [hunk 1]
```diff
@@ -109,6 +107,7 @@ var _ = SIGMigrationDescribe("Live Migration", func() {
 				Expect(resVMI.Status.EvacuationNodeName).To(Equal(""), "vmi evacuation state should be clean")
 			})
 
+			// Does it really makes sense?
```
Comment (jean-edouard) [inline tests/migration/eviction_strategy.go:110]: Please develop a little bit!

File: tests/migration/migration.go [hunk 2]
```diff
@@ -3125,8 +2904,7 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 
 			By("Deleting the Network Attachment Definition")
 			nad := libmigration.GenerateMigrationCNINetworkAttachmentDefinition()
-			err = virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(flags.KubeVirtInstallNamespace).Delete(context.TODO(), nad.Name, metav1.DeleteOptions{})
-			Expect(err).NotTo(HaveOccurred(), "Failed to delete the Network Attachment Definition")
+			Expect(virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(flags.KubeVirtInstallNamespace).Delete(context.TODO(), nad.Name, metav1.DeleteOptions{})).NotTo(HaveOccurred(), "Failed to delete the Network Attachment Definition")
```
Comment (jean-edouard) [inline tests/migration/migration.go:2907]:
```suggestion
			Expect(virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(flags.KubeVirtInstallNamespace).Delete(context.TODO(), nad.Name, metav1.DeleteOptions{})).To(Succeed(), "Failed to delete the Network Attachment Definition")
```

File: tests/migration/migration.go [hunk 3]
```diff
@@ -669,10 +605,11 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf(errorMessageFormat, pid, stdout, stderr, err))
 
 				// wait for both virtqemud to respawn and all connections to re-establish
+				// TODO flaky by design
```
Comment (jean-edouard) [inline tests/migration/migration.go:608]: Nit: you're not suggesting what **to do** about it :)

File: tests/migration/migration.go [hunk 4]
```diff
@@ -1754,20 +1624,11 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-							return nil
-						}
-
-						return fmt.Errorf("still waiting on target pod to complete, current phase is %s", pod.Status.Phase)
-					}, 10*time.Second, time.Second).Should(Succeed(), "Target pod should exit quickly after migration fails.")
+					var err error
+					vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					Eventually(matcher.ThisPodWith(vmi.Namespace, vmi.Status.MigrationState.TargetPod)).WithPolling(time.Second).WithTimeout(10*time.Second).
+						Should(Or(matcher.BeInPhase(k8sv1.PodFailed), matcher.BeInPhase(k8sv1.PodSucceeded)), "Target pod should exit quickly after migration fails.")
```
Comment (jean-edouard) [inline tests/migration/migration.go:1631]: Just to make the `Or` more explicit:
```suggestion
					Eventually(matcher.ThisPodWith(vmi.Namespace, vmi.Status.MigrationState.TargetPod)).WithPolling(time.Second).WithTimeout(10*time.Second).Should(
					Or(
						matcher.BeInPhase(k8sv1.PodFailed),
						matcher.BeInPhase(k8sv1.PodSucceeded)
					), "Target pod should exit quickly after migration fails.")
```

**Review by jean-edouard — COMMENTED** [review]
/approve
A few minor things in-line.
Also, do we have a document somewhere describing which tests qualify as conformance tests?
Really nice cleanup, thank you!

**Comment by xpivarc** [comment]
> /approve A few minor things in-line. Also, do we have a document somewhere describing which tests qualify as conformance tests? Really nice cleanup, thank you!

@jean-edouard  https://github.com/kubevirt/kubevirt/pull/12744

File: tests/migration/eviction_strategy.go (hunk 1 above)
Comment (fossedihelm) [reply tests/migration/eviction_strategy.go]: Each iteration waits for the pdb deletion before the next one.
So, basically, there is no correlation between each iteration.
But this was a note for me, and I forgot to delete :)

File: tests/migration/migration.go (hunk 4 above)
Comment (fossedihelm) [reply tests/migration/migration.go]: Done

File: tests/migration/migration.go (hunk 3 above)
Comment (fossedihelm) [reply tests/migration/migration.go]: Maybe something like this?
https://github.com/kubevirt/kubevirt/pull/12838/commits/27848c3403042fb94f82a6baab3112814bdfbc5a#diff-5034f7c82b377dd518dd147554ad7df560b0cc667f1f0aa08e7a632d1436237aR649-R652
WDYT?

**Comment by fossedihelm** [comment]
@jean-edouard Thank you for the review. PTAL

**Comment by fossedihelm** [comment]
> @fossedihelm the lint point to issue
> 
> Also I would like to review this as well, hopefully tomorrow

Sure, thank you! 
I am working on it :)

**Comment by fossedihelm** [comment]
> @fossedihelm failure are related to the changes.

Thank you! Should be fixed now

File: tests/migration/eviction_strategy.go [hunk 5]
```diff
@@ -353,7 +354,7 @@ var _ = SIGMigrationDescribe("Live Migration", func() {
 					expectVMIMigratedToAnotherNode(vmi.Name, node)
 				})
 
-				It("[test_id:2224] should handle mixture of VMs with different eviction strategies.", func() {
+				It("[test_id:2224] should handle mixture of VMs with different eviction strategies.", decorators.Conformance, func() {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply tests/migration/eviction_strategy.go:357]: ?

File: tests/migration/migration.go [hunk 6]
```diff
@@ -304,14 +304,7 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
 			It("[test_id:6969]should be successfully migrate with a tablet device", func() {
-				vmi := libvmifact.NewAlpineWithTestTooling(libnet.WithMasqueradeNetworking())
-				vmi.Spec.Domain.Devices.Inputs = []v1.Input{
-					{
-						Name: "tablet0",
-						Type: "tablet",
-						Bus:  "usb",
-					},
-				}
+				vmi := libvmifact.NewAlpineWithTestTooling(libnet.WithMasqueradeNetworking(), libvmi.WithInput("tablet0", v1.InputTypeTablet, v1.InputBusUSB))
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply tests/migration/migration.go:275]: Can I ask you why?

File: tests/migration/migration.go [hunk 7]
```diff
@@ -3421,3 +3404,20 @@ func haveMigrationState(matcher gomegatypes.GomegaMatcher) gomegatypes.GomegaMat
 		}),
 	}))
 }
+
+func withKernelBoot() libvmi.Option {
+	return func(vmi *v1.VirtualMachineInstance) {
+		kernelBootFirmware := utils.GetVMIKernelBootWithRandName().Spec.Domain.Firmware
+		if vmiFirmware := vmi.Spec.Domain.Firmware; vmiFirmware == nil {
+			vmiFirmware = kernelBootFirmware
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply tests/migration/migration.go:3414]: why?

[the remark this answers is not in the document]
Comment (xpivarc) [reply tests/migration/migration.go:3414]: You assign pointer to variable `vmiFirmware := vmi.Spec.Domain.Firmware`, and if it is nil then assign to the pointer (variable, not the struct) different pointer.  In other words this does not modify the vmi...

[the remark this answers is not in the document]
Comment (fossedihelm) [reply tests/migration/migration.go]: Oh! Nice one! You fixed a bug :)
Looks like it was there for a while https://github.com/kubevirt/kubevirt/blob/main/tests/migration/migration.go#L131-L140

[the remark this answers is not in the document]
Comment (fossedihelm) [reply tests/migration/migration.go]: Fixed

File: tests/migration/eviction_strategy.go [hunk 8]
```diff
@@ -72,7 +73,7 @@ var _ = SIGMigrationDescribe("Live Migration", func() {
 				vmi = alpineVMIWithEvictionStrategy()
 			})
 
-			It("[test_id:3242]should block the eviction api and migrate", func() {
+			It("[test_id:3242]should block the eviction api and migrate", decorators.Conformance, func() {
```
Comment (xpivarc) [inline tests/migration/eviction_strategy.go:76]: Lets remove the `BeforeEach`

Comment (xpivarc) [inline tests/migration/eviction_strategy.go:76]: While to some obvious, please reason in the commit message why you choose this test.

File: tests/migration/migration.go [hunk 9]
```diff
@@ -244,30 +244,6 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 	})
 
 	Describe("Starting a VirtualMachineInstance ", func() {
-		Context("with a bridge network interface", func() {
```
Comment (xpivarc) [inline tests/migration/migration.go:247]: Please mention the name (at least partial) in commit

**Review by xpivarc — COMMENTED** [review]
I made it to eb57c67c79bf6c67a964d5b900cfc0072c21496f. Can we split this PR to two with this commit as last?

File: tests/migration/migration.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (xpivarc) [reply tests/migration/migration.go:278]: You are asserting the values have not changed.

File: tests/migration/eviction_strategy.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (fossedihelm) [reply tests/migration/eviction_strategy.go:356]: OMG!

File: tests/migration/eviction_strategy.go (hunk 8 above)
Comment (fossedihelm) [reply tests/migration/eviction_strategy.go:70]: Added separate commit

Comment (fossedihelm) [reply tests/migration/eviction_strategy.go:70]: Done

File: tests/migration/migration.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (fossedihelm) [reply tests/migration/migration.go:278]: done

**Comment by fossedihelm** [comment]
> I made it to [eb57c67](https://github.com/kubevirt/kubevirt/commit/eb57c67c79bf6c67a964d5b900cfc0072c21496f). Can we split this PR to two with this commit as last?

@xpivarc Done

File: tests/migration/migration.go [hunk 10]
```diff
@@ -625,18 +627,16 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 				By("Checking that the VirtualMachineInstance console has expected output")
 				Expect(console.LoginToAlpine(vmi)).To(Succeed())
 
-				pods, err := virtClient.CoreV1().Pods(vmi.Namespace).List(context.Background(), metav1.ListOptions{
-					LabelSelector: v1.CreatedByLabel + "=" + string(vmi.GetUID()),
-				})
-				Expect(err).ToNot(HaveOccurred(), "Should list pods successfully")
-				Expect(pods.Items).To(HaveLen(1), "There should be only one VMI pod")
+				pod, err := libpod.GetPodByVirtualMachineInstance(vmi, vmi.Namespace)
```
Comment (xpivarc) [inline tests/migration/migration.go:630]: This is wrong as `GetPodByVirtualMachineInstance` always returns the new Pod. I would preserve the one Pod assert.

File: tests/migration/migration.go [hunk 11]
```diff
@@ -645,12 +645,11 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 				errorMessageFormat := "failed after running `kill -9 %v` with stdout:\n %v \n stderr:\n %v \n err: \n %v \n"
 				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf(errorMessageFormat, pid, stdout, stderr, err))
 
-				// wait for both virtqemud to respawn and all connections to re-establish
-				time.Sleep(30 * time.Second)
```
Comment (xpivarc) [inline tests/migration/migration.go:672]: +1000

File: tests/migration/migration.go [hunk 12]
```diff
@@ -645,12 +645,11 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-
 				// ensure new pid comes online
-				newPid := getVirtqemudPid(&pods.Items[0])
-				Expect(pid).ToNot(Equal(newPid), fmt.Sprintf("expected virtqemud to be cycled. original pid %s new pid %s", pid, newPid))
+				Eventually(getVirtqemudPid).WithArguments(pod).
+					WithPolling(5*time.Second).
```
Comment (xpivarc) [inline tests/migration/migration.go:650]: Can we be more aggressive than 5 seconds or is it not worth?

File: tests/migration/migration.go [hunk 13]
```diff
@@ -766,37 +755,14 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-						return err
-					}
-
-					if migration.Status.Phase == v1.MigrationFailed {
-						return fmt.Errorf("Migration should observe timeout period before transitioning to failed state")
-					}
-					return nil
-
-				}, 1*time.Minute, 10*time.Second).Should(Succeed())
+				Consistently(ThisMigration(migration)).WithPolling(10 * time.Second).WithTimeout(1 * time.Minute).Should(Not(BeInPhase(v1.MigrationFailed)))
```
Comment (xpivarc) [inline tests/migration/migration.go:758]: Let's use the error constant as additional description...

File: tests/migration/migration.go [hunk 14]
```diff
@@ -766,37 +755,14 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-					if err != nil {
-						return err
-					}
-
-					if migration.Status.Phase != v1.MigrationFailed {
-						return fmt.Errorf("Waiting on migration with phase %s to reach phase Failed", migration.Status.Phase)
-					}
-					return nil
-				}, 2*time.Minute, 5*time.Second).Should(Succeed(), "migration creation should fail")
+				Eventually(ThisMigration(migration)).WithPolling(5 * time.Second).WithTimeout(2 * time.Minute).Should(BeInPhase(v1.MigrationFailed))
```
Comment (xpivarc) [inline tests/migration/migration.go:761]: Let's use the error constant as additional description...

File: tests/migration/migration.go [hunk 15]
```diff
@@ -806,46 +772,21 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-						return err
-					}
-
-					if migration.Status.Phase == v1.MigrationFailed {
-						return fmt.Errorf("Migration should observe timeout period before transitioning to failed state")
-					}
-					return nil
-
-				}, 1*time.Minute, 10*time.Second).Should(Succeed())
+				Consistently(ThisMigration(migration)).WithPolling(10 * time.Second).WithTimeout(1 * time.Minute).Should(Not(BeInPhase(v1.MigrationFailed)))
```
Comment (xpivarc) [inline tests/migration/migration.go:786]: Let's use the error constant as additional description...

File: tests/migration/migration.go [hunk 16]
```diff
@@ -806,46 +772,21 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-					if err != nil {
-						return err
-					}
-
-					if migration.Status.Phase != v1.MigrationFailed {
-						return fmt.Errorf("Waiting on migration with phase %s to reach phase Failed", migration.Status.Phase)
-					}
-					return nil
-				}, 2*time.Minute, 5*time.Second).Should(Succeed(), "migration creation should fail")
+				Eventually(ThisMigration(migration)).WithPolling(5 * time.Second).WithTimeout(2 * time.Minute).Should(BeInPhase(v1.MigrationFailed))
```
Comment (xpivarc) [inline tests/migration/migration.go:789]: Let's use the error constant as additional description...

File: tests/migration/migration.go [hunk 17]
```diff
@@ -1730,21 +1661,11 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-						Expect(err).ToNot(HaveOccurred())
-
-						if pod.Status.Phase == k8sv1.PodFailed || pod.Status.Phase == k8sv1.PodSucceeded {
-							return nil
-						}
-
-						return fmt.Errorf("still waiting on target pod to complete, current phase is %s", pod.Status.Phase)
-					}, 10*time.Second, time.Second).Should(Succeed(), "Target pod should exit quickly after migration fails.")
+					var err error
+					vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
```
Comment (xpivarc) [inline tests/migration/migration.go:1665]: Do you need this?

File: tests/migration/migration.go [hunk 18]
```diff
@@ -1730,21 +1661,11 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
+					Expect(err).ToNot(HaveOccurred())
+					Eventually(ThisPodWith(vmi.Namespace, vmi.Status.MigrationState.TargetPod)).WithPolling(time.Second).WithTimeout(10*time.Second).
```
Comment (xpivarc) [inline tests/migration/migration.go:1667]: Beautiful.

File: tests/migration/migration.go [hunk 19]
```diff
@@ -1770,29 +1692,20 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-					Expect(err).ToNot(HaveOccurred())
-
-					if pod.Status.Phase == k8sv1.PodFailed || pod.Status.Phase == k8sv1.PodSucceeded {
-						return nil
-					}
-
-					return fmt.Errorf("still waiting on target pod to complete, current phase is %s", pod.Status.Phase)
-				}, 10*time.Second, time.Second).Should(Succeed(), "Target pod should exit quickly after migration fails.")
+				var err error
+				vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
```
Comment (xpivarc) [inline tests/migration/migration.go:1696]: Doesn't `ConfirmVMIPostMigrationFailed` return fresh vmi?

File: tests/migration/migration.go [hunk 20]
```diff
@@ -3334,11 +3171,6 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 
 				By("Checking that no backoff event occurred")
 				events.ExpectNoEvent(vmi, k8sv1.EventTypeWarning, controller.MigrationBackoffReason)
-				events, err := virtClient.CoreV1().Events(testsuite.NamespaceTestDefault).List(context.Background(), metav1.ListOptions{})
```
Comment (xpivarc) [inline tests/migration/migration.go:3351]: hah

**Review by xpivarc — COMMENTED** [review]
Thank you for the split!

File: tests/migration/migration.go (hunk 10 above)
Comment (fossedihelm) [reply tests/migration/migration.go]: I would argue that at this stage there will be only 1 pod or the `GetPodByVirtualMachineInstance` returns an error, but I don't insist

File: tests/migration/migration.go (hunk 12 above)
Comment (fossedihelm) [reply tests/migration/migration.go]: I think we can and we **must**. Currently, the timeout is only 10 seconds and hence there will be probably only 1 try.
I also increased the timeout, to be aligned with the previously 30 seconds of sleep.

File: tests/migration/migration.go (hunk 13 above)
Comment (fossedihelm) [reply tests/migration/migration.go:761]: I think the `BeInPhase` already provides a quite good description, no?

File: tests/migration/migration.go (hunk 17 above)
Comment (fossedihelm) [reply tests/migration/migration.go]: I don't think so

File: tests/migration/migration.go (hunk 19 above)
Comment (fossedihelm) [reply tests/migration/migration.go]: Yep +1

**Review by xpivarc — COMMENTED** [review]
/approve

**Comment by dasionov** [comment]
can we backport this refactor to 1.3 as well?

**Pull request merged**